### PR TITLE
feat: Record all calls

### PR DESCRIPTION
**Pull request by xquanluu**

### CODE REVIEW

**Comment by davehorton** [comment]
I get this warning when testing
```
3|jambonz-api-server      | (node:30196) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.
3|jambonz-api-server      | Please migrate your code to use AWS SDK for JavaScript (v3).
3|jambonz-api-server      | For more information, check the migration guide at https://a.co/7PzMCcy
3|jambonz-api-server      | (Use `node --trace-warnings ...` to show where the warning was created)
3|jambonz-api-server      | {"level":30, "time": "2023-05-31T18:24:27.924Z","pid":30196,"hostname":"ip-10-0-53-191","obj":{"sampleRate":8000,"mixType":"stereo","ca
```

We should update to AWS SDK v3

**Comment by xquanluu** [comment]
Hi @avoylenko 
The first working commit was using wav file, then we identified that the wav file was too big size, 1 min 30 seconds duration call could cost ~ 4MB file. after that we chose mp3 for this feature.

Hi @davehorton 
should we have new drop down list  in bucket credential for choosing which type of audio(wav/mp3) should be stored?

**Comment by davehorton** [comment]
yes, let's offer configuration - wav vs mp3

**Review by davehorton — APPROVED** [review]

**Review by davehorton — APPROVED** [review]

**Review by davehorton — APPROVED** [review]

File: db/upgrade-jambonz-db.js [hunk 1]
```diff
@@ -139,6 +139,11 @@ const sql = {
     'ALTER TABLE lcr_carrier_set_entry ADD FOREIGN KEY lcr_route_sid_idxfk (lcr_route_sid) REFERENCES lcr_routes (lcr_route_sid)',
     'ALTER TABLE lcr_carrier_set_entry ADD FOREIGN KEY voip_carrier_sid_idxfk_3 (voip_carrier_sid) REFERENCES voip_carriers (voip_carrier_sid)',
     'SET FOREIGN_KEY_CHECKS=1',
+  ],
+  8004: [
```
Comment (davehorton) [inline db/upgrade-jambonz-db.js:143]: please add the record_format alter table statement here so when we upgrade existing databases it is created

**Review by davehorton — CHANGES_REQUESTED** [review]

File: lib/record/s3.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+const wav = require('wav');
+
+async function upload(logger, socket) {
+
+  socket.on('message', async function(data, isBinary) {
+    try {
+      if (!isBinary) {
+        const obj = JSON.parse(data.toString());
+        logger.info({obj}, 'received JSON message from jambonz');
+        const {sampleRate, accountSid, callSid} = obj;
```
Comment (davehorton) [inline lib/record/s3.js:14]: In addition to these properties, I would also like to create metadata properties in the S3 uploaded file for:
- direction
- from
- to 
- callId
- applicationSid
- originatingSipIp
- originatingSipTrunkName

**Review by davehorton — CHANGES_REQUESTED** [review]

File: db/upgrade-jambonz-db.js (hunk 1 above)
Comment (davehorton) [reply db/upgrade-jambonz-db.js:143]: @xquanluu this is still needed

**Review by davehorton — CHANGES_REQUESTED** [review]

File: db/jambones-sql.sql [hunk 3]
```diff
@@ -506,6 +507,9 @@ subspace_client_secret VARCHAR(255),
 subspace_sip_teleport_id VARCHAR(255),
 subspace_sip_teleport_destinations VARCHAR(255),
 siprec_hook_sid CHAR(36),
+record_all_calls BOOLEAN NOT NULL DEFAULT false,
+record_format VARCHAR(16) NOT NULL DEFAULT 'mp3',
```
Comment (davehorton) [inline db/jambones-sql.sql:511]: this needs to be added to the upgrade db script

**Review by davehorton — CHANGES_REQUESTED** [review]

File: lib/routes/api/accounts.js [hunk 4]
```diff
@@ -675,6 +708,50 @@ account_subscriptions WHERE account_sid = ?)
...
+      const service_provider_sid = req.user.hasServiceProviderAuth ? req.user.service_provider_sid : null;
+      const results = await Account.retrieve(account_sid, service_provider_sid);
+      if (results.length === 0) throw new DbError('Invalid Account Sid');
+      const {bucket_credential} = results[0];
+      if (bucket_credential) {
+        const o = JSON.parse(decrypt(bucket_credential));
+        vendor = o.vendor;
+        switch (vendor) {
+          case 'aws_s3':
+            name = o.name;
```
Comment (davehorton) [inline lib/routes/api/accounts.js:733]: region is discarded if passed along, as a result this only works for us-east-1

**Review by davehorton — CHANGES_REQUESTED** [review]

**Review by davehorton — APPROVED** [review]

**Pull request merged**